agent/stats/task.go

```diff
@@ -0,0 +1,134 @@
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package stats
+
+import (
+	"context"
+	"fmt"
+	apitaskstatus "github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
+	"github.com/aws/amazon-ecs-agent/agent/stats/resolver"
+	"github.com/cihub/seelog"
+	"github.com/docker/docker/api/types"
+)
+
+
+
+type TaskStatsInterface interface {
+	GetTaskStats(task *StatsTask) (<-chan *types.StatsJSON, <-chan error)
+
+}
+
+func newStatsTaskContainer(taskARN string, containerPID string, numberOfContainers int,
+	resolver resolver.ContainerMetadataResolver,  client dockerapi.DockerClient) (*StatsTask, error) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+	return &StatsTask{
+		TaskMetadata: &TaskMetadata{
+			TaskArn:          taskARN,
+			ContainerPID:     containerPID,
+			NumberContainers: numberOfContainers,
+		},
+		Ctx:      ctx,
+		Cancel:   cancel,
+		Resolver: resolver,
+		client: client,
+	}, nil
+
+}
+
+func (task *StatsTask) StartStatsCollectionTask() {
+	queueSize := int(config.DefaultContainerMetricsPublishInterval.Seconds() * 4)
+	task.StatsQueue = NewQueue(queueSize)
+	task.StatsQueue.Reset()
+	go task.collect()
+}
+
+func (task *StatsTask) StopStatsCollectionTask() {
+	task.Cancel()
+}
+
+func (taskStat *StatsTask) collect() {
+	taskArn := taskStat.TaskMetadata.TaskArn
+	for {
+		select {
+		case <-taskStat.Ctx.Done():
+			seelog.Debugf("Stopping stats collection for taskStat %s", taskArn)
+			return
+		default:
+			err := taskStat.processStatsStream()
+			if err != nil {
+				seelog.Debugf("Error querying stats for task %s: %v", taskArn, err)
+			}
+			// We were disconnected from the stats stream.
+			// Check if the task is terminal. If it is, stop collecting metrics.
+			terminal, err := taskStat.terminal()
+			if err != nil {
+				// Error determining if the task is terminal. clean-up anyway.
+				seelog.Warnf("Error determining if the task %s is terminal, stopping stats collection: %v",
+					taskArn, err)
+				taskStat.StopStatsCollectionTask()
+			} else if terminal {
+				seelog.Infof("Task %s is terminal, stopping stats collection", taskArn)
+				taskStat.StopStatsCollectionTask()
+			}
+		}
+	}
+}
+
+func (taskStat *StatsTask) processStatsStream() error {
+	taskArn := taskStat.TaskMetadata.TaskArn
+	if len(taskStat.TaskMetadata.DeviceName) == 0 {
+		var err error
+		taskStat.TaskMetadata.DeviceName, err = taskStat.client.PopulateNIDeviceList(taskStat.TaskMetadata.ContainerPID)
+		if err != nil {
+			return err
+		}
+	}
+	dockerStats, errC := taskStat.client.GetAWSVPCNetworkStats(taskStat.TaskMetadata.DeviceName,
```

Comment:
Should this be called dockerStats as it is not really populated by docker?

Reply:
Will rename this to a more appropriate one.